src/_implementation/manual_walk.ts

```diff
@@ -0,0 +1,112 @@
+/**
+ * @license
+ * Copyright (c) 2019 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import { hasNativeDeclarativeShadowRoots } from './feature_detect.js';
+import { isTemplate, isElement, hasNoParentElement } from './util.js';
+
+/*
+ * Traverses the DOM to find all <template> elements with a `shadowroot`
+ * attribute and move their content into a ShadowRoot on their parent element.
+ *
+ * This processing is done bottom up so that when top-level <template>
+ * elements are hydrated, their contents are already hydrated and in the
+ * final correct structure of elements and shadow roots.
+ */
+export const hydrateShadowRoots = (root: ParentNode) => {
+  if (hasNativeDeclarativeShadowRoots()) {
+    return;  // nothing to do
+  }
+
+  // Approaches to try and benchmark:
+  //  - manual walk (current implementation)
+  //  - querySelectorAll
+  //  - TreeWalker
+
+  // Stack of nested templates that we're currently processing. Use to
+  // remember how to get from a <template>.content DocumentFragment back to
+  // its owner <template>
+  const templateStack: Array<HTMLTemplateElement> = [];
+
+  let currentNode: Element|DocumentFragment|null = root.firstElementChild;
+
+  // The outer loop traverses down, looking for <template shadowroot>
+  // elements. The inner loop traverses back up, hydrating them in a postorder
+  // traversal.
+  while (currentNode !== root && currentNode !== null) {
+    if (isTemplate(currentNode)) {
```

Comment:
Is this the right check? Don't we only want to push `<template shadowroot>`s onto the stack?

Reply:
No, because for correctness we need to hydrate shadow roots inside of templates.

Reply:
We don't do that for other elements though. That would require traversing every element and checking for a shadow root.